llmfoundry/models/layers/ffn.py

```diff
@@ -200,131 +225,225 @@ def build_ffn(
-            device=device,
-            bias=bias,
+    assert te is not None
+    ffn_hidden_size = resolve_ffn_hidden_size(d_model, expansion_ratio,
+                                              ffn_hidden_size)
+    if ffn_act_fn is not None:
+        raise ValueError(
+            f'Transformer Engine block does not support custom activation functions.'
         )
-    elif ffn_type == 'te_ln_mlp':
-        if te is None:
-            raise RuntimeError(
-                'Requirements for TransformerEngine not installed; see install instructions in `README.md`.'
-            )
-        ffn_hidden_size = resolve_ffn_hidden_size(d_model, expansion_ratio,
-                                                  ffn_hidden_size)
-        if ffn_act_fn is not None:
-            raise ValueError(
-                f'Transformer Engine block does not support custom activation functions.'
-            )
-        return te.LayerNormMLP(
-            hidden_size=d_model,
-            ffn_hidden_size=ffn_hidden_size,
-            bias=bias,
-            **kwargs,
-        )
-    elif ffn_type in ('mb_moe', 'mb_dmoe'):
-        if megablocks is None:
-            raise RuntimeError(
-                'Requirements for megablocks not installed; see install instructions in `README.md`.'
-            )
-        args = kwargs['args']
-        args.bias = bias
-        args.hidden_size = d_model
-        args.device = device
+    return te.LayerNormMLP(
+        hidden_size=d_model,
+        ffn_hidden_size=ffn_hidden_size,
+        bias=bias,
+        **kwargs,
+    )
+
+
+def build_torch_dmoe(
+    d_model: int,
+    expansion_ratio: Union[int, float],
+    ffn_hidden_size: Optional[int] = None,
+    ffn_act_fn: Optional[dict] = None,
+    device: Optional[str] = None,
+    bias: bool = True,
+    **kwargs: Any,
+) -> nn.Module:
+    moe_num_experts = kwargs.pop('moe_num_experts')
+    moe_top_k = kwargs.pop('moe_top_k')
+    mlp_type = kwargs.pop('mlp_type')
+    moe_jitter_eps = kwargs.pop('moe_jitter_eps')
+    moe_normalize_expert_weights = kwargs.pop('moe_normalize_expert_weights')
+    uniform_expert_assignment = kwargs.pop('uniform_expert_assignment')
+
+    fc_type = kwargs.pop('fc_type', 'torch')
+    del fc_type  # Unused
+
+    if len(kwargs) > 0:
+        raise ValueError(f'Invalid arguments to torch dmoe: {kwargs}.')
+
+    return dMoE(
+        hidden_size=d_model,
+        ffn_hidden_size=resolve_ffn_hidden_size(d_model, expansion_ratio,
+                                                ffn_hidden_size),
+        moe_num_experts=moe_num_experts,
+        moe_top_k=moe_top_k,
+        mlp_type=mlp_type,
+        bias=bias,
+        moe_jitter_eps=moe_jitter_eps,
+        activation_fn=resolve_ffn_act_fn(ffn_act_fn),
+        moe_normalize_expert_weights=moe_normalize_expert_weights,
+        uniform_expert_assignment=uniform_expert_assignment,
+        device=device,  # pyright: ignore[reportGeneralTypeIssues]
```

Comment:
Why ignore? I think we can remove this 🤔

Reply:
hm nvm I think @bigning has a PR for this

Reply:
Ah ok, I'll try. Was just a copy paste from existing code.

Reply:
ok fixed